sway-core/src/semantic_analysis/namespace/trait_map.rs

```diff
@@ -1352,42 +1354,67 @@ impl TraitMap {
         }
         let hash = hasher.finish();
 
-        if self.satisfied_cache.contains(&hash) {
-            return Ok(());
+        {
+            let trait_map = &mut module.current_lexical_scope_mut().items.implemented_traits;
+            if trait_map.satisfied_cache.contains(&hash) {
+                return Ok(());
+            }
         }
 
+        let all_impld_traits: BTreeSet<(Ident, TypeId)> =
+            Self::get_all_implemented_traits(module, type_id, engines);
+
         // Call the real implementation and cache when true
-        match self.check_if_trait_constraints_are_satisfied_for_type_inner(
+        match Self::check_if_trait_constraints_are_satisfied_for_type_inner(
             handler,
+            module,
             type_id,
             constraints,
             access_span,
             engines,
+            all_impld_traits,
             try_inserting_trait_impl_on_failure,
             code_block_first_pass,
         ) {
             Ok(()) => {
-                self.satisfied_cache.insert(hash);
+                let trait_map = &mut module.current_lexical_scope_mut().items.implemented_traits;
+                trait_map.satisfied_cache.insert(hash);
                 Ok(())
             }
             r => r,
         }
     }
 
-    #[allow(clippy::too_many_arguments)]
-    fn check_if_trait_constraints_are_satisfied_for_type_inner(
-        &mut self,
-        handler: &Handler,
+    fn get_all_implemented_traits(
+        module: &Module,
         type_id: TypeId,
-        constraints: &[TraitConstraint],
-        access_span: &Span,
         engines: &Engines,
-        try_inserting_trait_impl_on_failure: TryInsertingTraitImplOnFailure,
-        code_block_first_pass: CodeBlockFirstPass,
-    ) -> Result<(), ErrorEmitted> {
-        let type_engine = engines.te();
+    ) -> BTreeSet<(Ident, TypeId)> {
+        let mut lexical_scope_opt = Some(module.current_lexical_scope());
+        let mut all_impld_traits: BTreeSet<(Ident, TypeId)> = Default::default();
+        while let Some(lexical_scope) = lexical_scope_opt {
+            all_impld_traits.extend(
+                lexical_scope
+                    .items
+                    .implemented_traits
+                    .get_implemented_traits(type_id, engines),
+            );
```

Comment:
I suppose my question really is: Is there a case where this algorithm will throw an ambiguity error, but where shadowing actually resolves the ambiguity? I'm not sure, because I'm not sure a trait implementation can shadow another trait implementation.